EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,439 @@
+
+The domain separator prevents collision of otherwise identical structures. It is possible that two DApps come up with an identical structure like `Transfer(address from,address to,uint256 amount)` that should not be compatible. By introducing a domain separator the DApp developers are guaranteed that there can be no signature collision.
+
+The domain separator also allows for multiple distinct signatures use-cases on the same struct instance within a given DApp. In the previous example, perhaps signatures from both `from` and `to` are required. By providing two distinct domain separators these signatures can be distinguished from each other.
+
+**Alternative 1**: Use the target contract address as domain separator. This solves the first problem, contracts coming up with identical types, but does not address second use-case. The standard does suggest implementors to use the target contract address where this is appropriate.
+
+The function `hashStruct` starts with a `typeHash` to separate types. By giving different types a different prefix the `encodeData` function only has to be injective for within a given type. It is okay for `encodeData(a)` to equal `encodeData(b)` as long as `typeOf(a)` is not `typeOf(b)`.
+
+### Rationale for `typeHash`
+
+The `typeHash` is designed to turn into a compile time constant in Solidity. For example:
+
+```javascript
+bytes32 constant MESSAGE_TYPEHASH = keccak256(
+  "Message(address from,address to,string contents)");
+```
+
+For the type hash several alternatives where considered and rejected for the reasons:
+
+**Alternative 2**: Use ABIv2 function signatures. `bytes4` is not enough to be collision resistant. Unlike function signatures, there is negligible runtime cost incurred by using longer hashes.
+
+**Alternative 3**: ABIv2 function signatures modified to be 256-bit. While this captures type info, it does not capture any of the semantics other than the function. This is already causing a practical collision between ERC20's and ERC721's `transfer(address,uint256)`, where in the former the `uint256` revers to an amount and the latter to a unique id. In general ABIv2 favors compatibility where a hashing standard should prefer incompatibility.
+
+**Alternative 4**: 256-bit ABIv2 signatures extended with parameter names and struct names. The `Message` example from a above would be encoded as `Message(Person(string name,address wallet) from,Person(string name,address wallet) to,string message)`. This is longer than the proposed solution. And indeed, the length of the string can grow exponentially in the length of the input (consider `struct A{B a;B b;}; struct B {C a;C b;}; …`). It also does not allow a recursive struct type (consider `struct List {uint256 value; List next;}`).
```

Comment:
Since Solidity doesn't permit recursive structs, this seems a bit academic, and since you're hashing the result, length seems like a nonissue too.

Reply:
Correct. But since there is basically no difference in implementation complexity, I think its best to pick the most generic option. This way the standard does not have to be changed when a future version of Solidity adds support for recursive structs.

Reply:
Struct fields in other structs are generally treated as value types, so I think a recursive struct is outright impossible.